### DISCUSSION TITLE
Incompatible types in assignment for Enum

### DISCUSSION THREAD

**Discussion by miketheman** [sqlalchemy/sqlalchemy#9941] [Usage Questions] [answered]
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [X] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Describe the typing issue

Using a custom enum, mypy raises an error to annotate a column

### To Reproduce

```python
import enum

from sqlalchemy import Column, Enum
from sqlalchemy.orm import declarative_base


ModelBase = declarative_base()

# also tried with:
# from sqlalchemy.orm import DeclarativeBase
# class ModelBase(DeclarativeBase):
#    ...
# Same result.


class BanReason(enum.Enum):
    AUTHENTICATION_ATTEMPTS = "authentication-attempts"


class IpAddress(ModelBase):
    ban_reason = Column(Enum(BanReason), nullable=True)


ip = IpAddress()
ip.ban_reason = BanReason.AUTHENTICATION_ATTEMPTS

reveal_type(ip.ban_reason)
```

### Error

```shellsession
$ mypy repro.py
repro.py:22: error: Need type annotation for "ban_reason"  [var-annotated]
repro.py:25: error: Incompatible types in assignment (expression has type "BanReason", variable has type "Column[Any]")  [assignment]
repro.py:27: note: Revealed type is "sqlalchemy.sql.schema.Column[Any]"
Found 2 errors in 1 file (checked 1 source file)
``` 

### Versions

- OS: macOS, Linux Debian bullseye
- Python: 3.11.2, 3.11.3
- SQLAlchemy: 2.016
- Type checker (eg: mypy 0.991, pyright 1.1.290, etc): mypy==1.0.1 mypy-extensions==1.0.0 mypy-zope==0.9.1

### Additional context

More full-featured stack and more type-related error messages can be found on https://github.com/pypi/warehouse/pull/13929 (specifically https://github.com/pypi/warehouse/actions/runs/5246828369/jobs/9476223258?pr=13929 )

I've tried adding annotations to the enum:

```python
class BanReason(str, enum.Enum):
    AUTHENTICATION_ATTEMPTS = "authentication-attempts"
```
Same result.

Adding annotation directly to the Column:

```python
class IpAddress(ModelBase):
    ban_reason: str = Column(Enum(BanReason), nullable=True)
```
(`str` could be replaced with anything here)
Error mesasage:
```
repro.py:22: error: Incompatible types in assignment (expression has type "Column[<nothing>]", variable has type "str")  [assignment]
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

All those are expected. Mypy is supported using the new style declarative. I'm not 100% sure what's accepted by the mypy plugin, but that's deprecated.

Please see the documentation regarding the migration: https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#orm-declarative-models

**Reply by miketheman** [reply]
Thanks @CaselIT for the pointer! That makes a lot of sense - I must have overlooked it from the migration guide.

**Reply by CaselIT** [reply]
Well the new mypy support is in the what's new document. The 1.4 style work perfectly fije in v2 but it's not supported by type checkers

**Reply by miketheman** [reply]
During the conversion to 2.0 declarative style top support the new type checking I'm hitting a confusing issue with a specific use case.

In 1.4 we use AbstractConcreteBase to create an `Event`, and then a mixin called `HasEvents` that declares `__init_subclass__()` to add a named subclass dynamically to the parent.
What we want is to have concrete tables for the classes mixing in (and relationships to the events).
For example, `User(HasEvents)` should end up with `user` and `user_events` tables, and we refer to `User.Event` and `User.events`

[This example](https://docs.sqlalchemy.org/en/20/_modules/examples/generic_associations/table_per_related.html) seems to be close, but not exactly, as it doesn't create an abstract class I can hook onto.

Is there some other part of the docs I'm missing on how to compose this kind of behavior in 2.0 DeclarativeBase style you can point me to?

**Reply by CaselIT** [reply]
I think that should work correctly also on 2.0. Is the issue a typing one or is it at runtime?
Do you have an example you can point to so that's clearer what you are trying to do?

**Reply by miketheman** [reply]
I'm in the middle of shipping the 2.0 upgrade with 1.4 `declarative_base` still intact, and will work up an example PR after it's merged that should display the issue. It might be a bit complex - I'm not entirely sure where to start carving it apart to make it a clear example.

Should that be a different Discussion, or should I rename/reshape this one?

**Reply by CaselIT** [reply]
I think a different discussion may be best
